test/functional/horizon_controller_test.go

```diff
@@ -0,0 +1,279 @@
+}
+
+// Gets the condition of given type from the resource.
+func (t TestHorizon) GetCondition(conditionType condition.Type, reason condition.Reason) condition.Condition {
+	t.Refresh()
+	if t.Instance.Status.Conditions == nil {
+		return condition.Condition{}
+	}
+
+	cond := t.Instance.Status.Conditions.Get(conditionType)
+	if cond != nil && cond.Reason == reason {
+		return *cond
+	}
+
+	return condition.Condition{}
+
+}
+
+type TestKeystoneAPI struct {
+	LookupKey types.NamespacedName
+	// The input data for creating a KeystoneAPI
+	Template *keystonev1.KeystoneAPI
+	Instance *keystonev1.KeystoneAPI
+}
+
+// NewTestKeystoneAPI initializes the the input for a KeystoneAPI instance
+// but does not create it yet. So the client can finetuned the Template data
+// before calling Create()
+func NewTestKeystoneAPI(namespace string) *TestKeystoneAPI {
+	name := fmt.Sprintf("keystone-%s", uuid.New().String())
+	template := &keystonev1.KeystoneAPI{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "keystone.openstack.org/v1beta1",
+			Kind:       "KeystoneAPI",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: TestNamespace,
+		},
+		Spec: keystonev1.KeystoneAPISpec{
+			DatabaseUser: "foo-bar-baz",
+		},
+		Status: keystonev1.KeystoneAPIStatus{
+			APIEndpoints: map[string]string{
+				"internal": "fake-keystone-internal-endpoint",
+				"public":   "fake-keystone-public-endpoint",
+			},
+			DatabaseHostname: "fake-database-hostname",
+		},
+	}
+	return &TestKeystoneAPI{
+		LookupKey: types.NamespacedName{Name: name, Namespace: namespace},
+		Template:  template,
+		Instance:  &keystonev1.KeystoneAPI{},
+	}
+}
+
+// Creates the KeystoneAPI resource in k8s based on the Template. The Template
+// is not updated with the result of the create.
+func (t TestKeystoneAPI) Create() {
+	t.Instance = t.Template.DeepCopy()
+	Expect(k8sClient.Create(ctx, t.Instance)).Should(Succeed())
+
+	// the Status field needs to be written via a separate client
+	t.Instance.Status = t.Template.Status
+	Expect(k8sClient.Status().Update(ctx, t.Instance)).Should(Succeed())
+}
+
+// Deletes the KeystoneAPI resource from k8s and waits until it is deleted.
+func (t TestKeystoneAPI) Delete() {
+	Expect(k8sClient.Delete(ctx, t.Template.DeepCopy())).Should(Succeed())
+	// We have to wait for the instance to be fully deleted
+	// by the controller
+	Eventually(func() bool {
+		err := k8sClient.Get(ctx, t.LookupKey, t.Instance)
+		return k8s_errors.IsNotFound(err)
+	}, timeout, interval).Should(BeTrue())
+}
+
+var _ = Describe("Horizon controller", func() {
+
+	var horizon TestHorizon
+	var secret *corev1.Secret
+	var keystoneAPI *TestKeystoneAPI
+
+	BeforeEach(func() {
+		// lib-common uses OPERATOR_TEMPLATES env var to locate the "templates"
+		// directory of the operator. We need to set them othervise lib-common
+		// will fail to generate the ConfigMap as it does not find common.sh
+		err := os.Setenv("OPERATOR_TEMPLATES", "../../templates")
```

Comment:
you don't need this if you add it this env var to gotest make target as well similarly how you added it to the make test target